lychee-lib/src/uri.rs

```diff
@@ -15,30 +16,56 @@ pub enum Uri {
 }
 
 impl Uri {
+    #[must_use]
     pub fn as_str(&self) -> &str {
         match self {
-            Uri::Website(url) => url.as_str(),
-            Uri::Mail(address) => address.as_str(),
+            Uri::Website(url) => url.as_ref(),
+            Uri::Mail(address) => address.as_ref(),
         }
     }
 
+    #[must_use]
     pub fn scheme(&self) -> Option<String> {
-        match self {
-            Uri::Website(url) => Some(url.scheme().to_string()),
-            Uri::Mail(_address) => None,
+        if let Uri::Website(url) = self {
+            Some(url.scheme().to_string())
+        } else {
+            None
         }
     }
 
+    #[must_use]
     pub fn host_ip(&self) -> Option<IpAddr> {
         match self {
             Self::Website(url) => match url.host()? {
                 url::Host::Ipv4(v4_addr) => Some(v4_addr.into()),
                 url::Host::Ipv6(v6_addr) => Some(v6_addr.into()),
-                _ => None,
+                url::Host::Domain(_) => None,
             },
             Self::Mail(_) => None,
         }
     }
+
+    pub(crate) fn extract_github(&self) -> Option<(&str, &str)> {
+        debug_assert!(
+            matches!(self, Uri::Website(_)),
+            "Should only be called on a Website type!"
+        );
+        let url = if let Uri::Website(url) = self {
+            url
+        } else {
+            unsafe { std::hint::unreachable_unchecked() }
```

Comment:
Just wanted to mention that I still have mixed feelings about the use of unsafe. 😄 
It's fine and we can merge it like that, but I don't see the need for introducing unsafe if the perf improvement is not significant and there are no other significant upsides. It might be beneficial, but we'd have to measure that. 
Yes, I'm aware that this totally safe in this case with the `debug_assert`, but why introduce unsafe when we can have a fully "safe" crate with 99% of the upsides? 
My main concern is about future refactorings and missing an edge-case. 
Sorry, don't want to start another long discussion. It's just a note that we have to revisit that topic and we might revert that decision in the future. 😅

Reply:
Well for this particular case, I'm able to avoid writing any explicit `unwrap`, `expect`, `unrechable` or `std::hint::unreachable_unchecked`.

```rust
        if let Uri::Website(url) = self {
            if matches!(url.domain(), Some("github.com") | Some("www.github.com")) {
                if let Some(mut path) = url.path_segments() {
                    if let Some(owner) = path.next() {
                        if let Some(repo) = path.next() {
                            return Some((owner, repo));
                        }
                    }
                }
            }
        }

```

Reply:
Extracting this in a separate `fn extract_github_info(url: &_) -> Option<(String, String)> { .. }` would allow you to simply use the `?` operator (`Option<_>` impls `Try`) and there would not need to be a single nesting layer :smiley:

Reply:
Good idea @drahnr. Would you be willing to open a PR for that?

Reply:
Actually, it was already implemented as `extract_github` in a follow up which I missed

Reply:
All good then. 👍